### PR TITLE
feat: support modern.js data action

### PR DESCRIPTION
**Pull request by yimingjfe**
## Summary

This PR implements the RFC https://github.com/web-infra-dev/modern.js/discussions/4755.

### <samp>🤖 Generated by Copilot at d28833c</samp>

This pull request adds support for data loader actions, which are optional files that define custom logic for fetching data in the `@modern-js/runtime` module. It refactors and updates the CLI, the runtime, and the app-tools modules to handle both loader and action requests, and to generate the appropriate code and types. It also adds and fixes some integration tests for the data loader feature, using different modes and scenarios.

### <samp>🤖 Generated by Copilot at d28833c</samp>

*  Add support for data loader actions in CLI, runtime, and app-tools modules
  * Create a new function `createActionRequest` that creates a request handler for data loader actions ([link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-11d382118af0b9e02cea623de0313c26c484e99b59379f3dca8d29f2fba68855R74-R115))
  * Modify the `generateClient` function to take an object with inline, action, and routeId properties and generate client code for data loader requests accordingly ([link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-dd189ac85d6f17d3d8bab015abdf52b55887f05aa545e70c51bd25207a973aabL4-R12), [link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-dd189ac85d6f17d3d8bab015abdf52b55887f05aa545e70c51bd25207a973aabL27-R36), [link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-dd189ac85d6f17d3d8bab015abdf52b55887f05aa545e70c51bd25207a973aabL57))
  * Modify the type `Context` to add more properties for data loader requests and parse them from the resourceQuery ([link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-8b1efb4fcc540bd78f483287d73e26762c2f2d274570d1d37b5ae621d18faeeeL8-R12), [link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-8b1efb4fcc540bd78f483287d73e26762c2f2d274570d1d37b5ae621d18faeeeL30-R42))
  * Modify the call to `generateClient` to pass the inline, action, and routeId properties from the options object ([link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-8b1efb4fcc540bd78f483287d73e26762c2f2d274570d1d37b5ae621d18faeeeL66-R69))
  * Rename the function `createLoaderRequest` to `createRequest` and modify it to handle both loader and action requests ([link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-f9da7a857e3fdfbcd54347f5d566dd980f03b59099c61ec0f38941505cf0b4feL82-R82), [link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-f9da7a857e3fdfbcd54347f5d566dd980f03b59099c61ec0f38941505cf0b4feL89-R91), [link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-f9da7a857e3fdfbcd54347f5d566dd980f03b59099c61ec0f38941505cf0b4feL95-R103), [link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-f9da7a857e3fdfbcd54347f5d566dd980f03b59099c61ec0f38941505cf0b4feL125-R133), [link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-f9da7a857e3fdfbcd54347f5d566dd980f03b59099c61ec0f38941505cf0b4feL134-L137), [link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-f9da7a857e3fdfbcd54347f5d566dd980f03b59099c61ec0f38941505cf0b4feL145-R150), [link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-f9da7a857e3fdfbcd54347f5d566dd980f03b59099c61ec0f38941505cf0b4feL190-R200), [link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-f9da7a857e3fdfbcd54347f5d566dd980f03b59099c61ec0f38941505cf0b4feL201))
  * Add a new constant `ACTION_EXPORT_NAME` that represents the name of the action export from the data loader files ([link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-b02a3b2a5a6c983ffc4a0be8f2783d9fd7eb47f3d82d9f66edf48494d56e7085R15-R16))
  * Modify the `walk` function to check if a data loader file has an action export and set the route.action property accordingly ([link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-89756710513f64f72aaf997fdd2a0179fd26970e9d18e579b4648498933ce1cdL5-R5), [link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-89756710513f64f72aaf997fdd2a0179fd26970e9d18e579b4648498933ce1cdL155-R164), [link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-89756710513f64f72aaf997fdd2a0179fd26970e9d18e579b4648498933ce1cdR209-R211), [link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-89756710513f64f72aaf997fdd2a0179fd26970e9d18e579b4648498933ce1cdL222-R236), [link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-89756710513f64f72aaf997fdd2a0179fd26970e9d18e579b4648498933ce1cdR260-R268), [link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-89756710513f64f72aaf997fdd2a0179fd26970e9d18e579b4648498933ce1cdR288-R290), [link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-89756710513f64f72aaf997fdd2a0179fd26970e9d18e579b4648498933ce1cdR317-R319))
  * Modify the `generateRoutes` and `generateLoaders` functions to include the action property in the route objects and the loadersMap object, and to use the loaderId, route, and action properties to generate the client code for data loader requests ([link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-dbfb9b73a9abfe46cea1f0f8a12245ca44db2fa24d49d3641c8e1ce883d3fa44R126-R137), [link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-dbfb9b73a9abfe46cea1f0f8a12245ca44db2fa24d49d3641c8e1ce883d3fa44L132-R163), [link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-dbfb9b73a9abfe46cea1f0f8a12245ca44db2fa24d49d3641c8e1ce883d3fa44R170), [link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-dbfb9b73a9abfe46cea1f0f8a12245ca44db2fa24d49d3641c8e1ce883d3fa44L153-R187), [link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-dbfb9b73a9abfe46cea1f0f8a12245ca44db2fa24d49d3641c8e1ce883d3fa44L164-R210), [link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-dbfb9b73a9abfe46cea1f0f8a12245ca44db2fa24d49d3641c8e1ce883d3fa44L204-R249), [link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-dbfb9b73a9abfe46cea1f0f8a12245ca44db2fa24d49d3641c8e1ce883d3fa44L225-R280), [link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-dbfb9b73a9abfe46cea1f0f8a12245ca44db2fa24d49d3641c8e1ce883d3fa44L231-R287), [link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-dbfb9b73a9abfe46cea1f0f8a12245ca44db2fa24d49d3641c8e1ce883d3fa44R300), [link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-dbfb9b73a9abfe46cea1f0f8a12245ca44db2fa24d49d3641c8e1ce883d3fa44L265-R330), [link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-dbfb9b73a9abfe46cea1f0f8a12245ca44db2fa24d49d3641c8e1ce883d3fa44R372), [link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-dbfb9b73a9abfe46cea1f0f8a12245ca44db2fa24d49d3641c8e1ce883d3fa44R394), [link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-dbfb9b73a9abfe46cea1f0f8a12245ca44db2fa24d49d3641c8e1ce883d3fa44L383-R475))
  * Add a new function `hasAction` that checks if a data loader file has an action export or not ([link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-0079e528411b9e11c7e8393a72120660449938bf82a9404088bbc310e3922a00R15), [link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-0079e528411b9e11c7e8393a72120660449938bf82a9404088bbc310e3922a00L148-R178))
  * Add a new test file `data.test.ts` that tests the `hasLoader` and `hasAction` functions ([link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-e886270b8ae5af133b5d8ef9430a562f0d8ad9b5d6dac8684cdac5cb7d7497a4R1-R22))
  * Add a new data file `$.data.ts` that exports a loader and an action function as named exports ([link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-2f86398366fa022471bd091d4d6713e526b281555eb82f9ac193e9e9f0af4061R1-R7))
  * Add a new file `utils.ts` that exports a constant `modernTestActionName` that represents the name of the action used in the integration tests ([link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-b6b3362846cf28440602ff30dcb2db4fd3a7662f9e78ceb70ade30c2787037dfR1))
  * Add a new data file `page.data.ts` in the user/[id] directory that exports a loader and an action function as named exports, and use them in the `page.tsx` file in the same directory ([link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-a1fbb984ac5ec352e17a14dba34727d80c637e11545b053f6118bd503730fadcR1-R14), [link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-5064f2d23f348801c654da6c22174bf36482d680aee59fe7b3a321073d0b109dL1-R5), [link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-5064f2d23f348801c654da6c22174bf36482d680aee59fe7b3a321073d0b109dL7-R25))
  * Add a new data file `page.data.ts` in the user/[id]/profile directory that exports a loader and an action function as named exports, and use them in the `page.tsx` file in the same directory ([link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-068002cfed0845449c8eaaa02be4b0c52c6e60385272796bef5dfb42316530b4R1-R21), [link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-465c62db76975353559c8a3e6755f256c37cb19a4fa84bd0d6d41aa408e59132L1-R5), [link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-465c62db76975353559c8a3e6755f256c37cb19a4fa84bd0d6d41aa408e59132L7-R34))
  * Add a new data file `page.data.ts` in the user/profile directory that exports a loader function as a named export ([link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-a65198ba8796b4feee473a2d64650a6d298d1bf304148f03828a55c899cf9f6bR1-R3))
  * Add two new functions `supportActionInCSR` and `supportActionInSSR` that test the data loader actions in CSR and SSR modes respectively, and use them in the test cases 'support action in CSR' and 'support action in SSR' in the `index.test.ts` file ([link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-a706802b24849d0e19f24d8b93af40f32a74278238ca698f1a17f1f8fc364444R521-R552), [link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-a706802b24849d0e19f24d8b93af40f32a74278238ca698f1a17f1f8fc364444R654-R663), [link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-a706802b24849d0e19f24d8b93af40f32a74278238ca698f1a17f1f8fc364444R775-R783), [link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-a706802b24849d0e19f24d8b93af40f32a74278238ca698f1a17f1f8fc364444R898-R906))
  * Modify the `supportNoLayoutDir` function to expect the correct text for the user/[id] route ([link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-a706802b24849d0e19f24d8b93af40f32a74278238ca698f1a17f1f8fc364444L217-R217))
  * Remove the commented out code that creates a new page ([link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-a706802b24849d0e19f24d8b93af40f32a74278238ca698f1a17f1f8fc364444L374))
  * Remove the skip annotations from the test cases 'support handle loader error' and 'support loader for ssr and csr', as they are fixed by the pull request ([link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-a706802b24849d0e19f24d8b93af40f32a74278238ca698f1a17f1f8fc364444L582-R612), [link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-a706802b24849d0e19f24d8b93af40f32a74278238ca698f1a17f1f8fc364444L596-R626), [link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-a706802b24849d0e19f24d8b93af40f32a74278238ca698f1a17f1f8fc364444L691-R731), [link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-a706802b24849d0e19f24d8b93af40f32a74278238ca698f1a17f1f8fc364444L807-R856), [link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-a706802b24849d0e19f24d8b93af40f32a74278238ca698f1a17f1f8fc364444L821-R870))
* Delete the unused file `page.loader.ts` in the user/profile directory ([link](https://github.com/web-infra-dev/modern.js/pull/4736/files?diff=unified&w=0#diff-d21d29592eca0a8d4a41c7e4c75036890476738d6e6a9752666cf71449a5f51f))

## Checklist

- [x] I have added changeset via `pnpm run change`.
- [x] I have updated the documentation.
- [x] I have added tests to cover my changes.

### CODE REVIEW

File: packages/cli/plugin-data-loader/src/runtime/index.ts [hunk 1]
```diff
@@ -79,20 +79,28 @@ const createLoaderHeaders = (
...
+  const init: {
+    [key: string]: unknown;
+  } = {
     method: context.method,
     headers: createLoaderHeaders(context.headers),
     signal: controller.signal,
   };
 
-  return new Request(url.href, init);
+  if (context.method.toUpperCase() !== 'GET' && context.method !== 'HEAD') {
```
Comment (GiveMe-A-Name) [inline packages/cli/plugin-data-loader/src/runtime/index.ts:97]: Is it better like this `!['GET', 'HEAD'].include(context.method.toUpperCase())`

File: tests/integration/routes/tests/index.test.ts [hunk 2]
```diff
@@ -579,7 +609,7 @@ describe('dev', () => {
       supportHandleConfig(page, appPort));
 
     // FIXME: skip the test
```
Comment (GiveMe-A-Name) [inline tests/integration/routes/tests/index.test.ts:611]: You can remove this remark.

File: tests/integration/routes/tests/index.test.ts [hunk 3]
```diff
@@ -688,7 +728,7 @@ describe('build', () => {
       supportPathWithoutLayout(page, errors, appPort));
 
     // FIXME: skip the test
```
Comment (GiveMe-A-Name) [inline tests/integration/routes/tests/index.test.ts:730]: also

Comment (GiveMe-A-Name) [inline tests/integration/routes/tests/index.test.ts:730]: also

Comment (yimingjfe) [reply tests/integration/routes/tests/index.test.ts]: done~

**Review by GiveMe-A-Name — APPROVED** [review]

**Pull request merged**